File: Editorial/0102_Binary_Tree_Level_Order_Traversal/C/2_Iteration.c
int** levelOrder(struct TreeNode* root, int* returnSize,
                 int** returnColumnSizes) {
    *returnSize = 0;
    if (!root) return NULL;
    typedef struct Node {
        void* data;
        struct Node* next;
    } Node;
    typedef struct Queue {
        Node* head;
        Node* tail;
    } Queue;
    Queue* createQueue() {
        Queue* queue = malloc(sizeof(Queue));
        queue->head = queue->tail = NULL;
        return queue;
    }
    void push(Queue * queue, void* data) {
        Node* node = malloc(sizeof(Node));
        node->data = data;
        node->next = NULL;
        if (queue->head == NULL) {
            queue->head = queue->tail = node;
        } else {
            queue->tail = queue->tail->next = node;
        }
    }
    void* pop(Queue * queue) {
        if (!queue->head) return NULL;
        Node* head = queue->head;
        void* data = head->data;
        queue->head = queue->head->next;
        if (queue->head == NULL) queue->tail = NULL;
        free(head);
        return data;
    }
    size_t size(Queue * queue) {
        size_t count = 0;
        for (Node* node = queue->head; node; node = node->next) count++;
        return count;
    }
    int isEmpty(Queue * queue) { return queue->head == NULL; }
    Queue* queue = createQueue();
    push(queue, root);
    int** levels = malloc(sizeof(int*) * 2000);
    *returnColumnSizes = (int*)malloc(sizeof(int) * 2000);
    while (!isEmpty(queue)) {
        int level_size = size(queue);
        levels[*returnSize] = malloc(sizeof(int) * level_size);
        (*returnColumnSizes)[*returnSize] = level_size;
        for (int i = 0; i < level_size; i++) {
            struct TreeNode* node = pop(queue);
            levels[*returnSize][i] = node->val;
            if (node->left) push(queue, node->left);
            if (node->right) push(queue, node->right);
        }
        (*returnSize)++;
    }
    return levels;
}

File: Editorial/0102_Binary_Tree_Level_Order_Traversal/Go/2_Iteration.go
import llq "github.com/emirpasic/gods/queues/linkedlistqueue"

func levelOrder(root *TreeNode) [][]int {
    var levels [][]int
    if root == nil {
        return levels
    }

    queue := llq.New()  // Initialize a new queue
    queue.Enqueue(root) // Start with the root node

    for !queue.Empty() {
        levelSize := queue.Size()
        currentLevel := make([]int, 0, levelSize)

        for i := 0; i < levelSize; i++ {
            node, _ := queue.Dequeue() // Dequeue returns (interface{}, bool)
            treeNode := node.(*TreeNode)

            currentLevel = append(
                currentLevel,
                treeNode.Val,
            ) // Fulfill the current level

            if treeNode.Left != nil {
                queue.Enqueue(treeNode.Left) // Enqueue left child
            }
            if treeNode.Right != nil {
                queue.Enqueue(treeNode.Right) // Enqueue right child
            }
        }

        levels = append(
            levels,
            currentLevel,
        ) // Append the current level to the results
    }

    return levels
}
